### ISSUE TITLE
SAASBO without BO

### ISSUE THREAD

**Issue by jrojsel** [facebook/Ax#697]
Hello,

I have been using BoTorch and GPyTorch for years, but I have never used Ax. Since the Sparse Axis-Aligned Subsplace (SAAS) model is not implemented in those libraries, I was hoping that there is a way to use the Ax implementation of the SAAS model to do inference. That is, I am not interested in doing BO, I only want to create a simple single output model and assess it by using the built-in cross-validation functionality, among other things. The SAASBO tutorial is not very clear on how to create+train the model using a user-supplied dataset.

My dataset consists of approximately 100 30-dimensional datapoints.

Could you please provide me with a minimal working example on how to do this?

Br,
Jimmy

**Comment by lena-kashtelyan** [comment]
cc @dme65, @sdaulton for help with this

**Comment by Balandat** [comment]
@lena-kashtelyan I actually think this isn't a SAASBO question inasmuch as it is a question about setting up custom arms and returning data to instantiate an Ax model. We should probably make a step by step tutorial showcasing how this is done so we can point to that.

@jrojsel, you'll want to set up an experiment and generate a bunch of custom trials that corresponds to your data points. 

So you'd define the search space and experiment as done in the tutorial, instead of a `BraninMetric` you just use a regular `Metric`.

Then you can generate custom trials like so
```
trials = []
for i, params in enumerate(custom_params):
    trial = experiment.new_trial(GeneratorRun(arms=[Arm(params, name=f"arm_{i}")]))
    trial.mark_running(no_runner_required=True)
    trials.append(trial)
```
where `custom_params` is an iterable of parameter dictionaries mapping the parameter name to the feature value. 

Then you can generate a dataframe with the trial results and make a `Data` object via `data=Data(df)`, where `df` is a pandas dataframe with columns `{"arm_name", "metric_name", "mean", "sem"}` (where `sem` can be `NaN` if you want to infer a noise level).

Now you can instantiate the model as in the tutorial via 
```
    model = Models.FULLYBAYESIAN(
        experiment=experiment, 
        data=data,
        num_samples=256,  # Increasing this may result in better model fits
        warmup_steps=512,  # Increasing this may result in better model fits
        gp_kernel="rbf",  # "rbf" is the default in the paper, but we also support "matern"
        torch_device=tkwargs["device"],
        torch_dtype=tkwargs["dtype"],
        verbose=False,  # Set to True to print stats from MCMC
        disable_progbar=True,  # Set to False to print a progress bar from MCMC
    )
```
and use the cross validation / prediction functionality as usual.

**Comment by lena-kashtelyan** [comment]
Oh my bad, thank you for jumping in @Balandat! 

@jrojsel, just FYI, you can also use the Service API to make the initial experiment using the `attach_trial` method (discussed in [this section](https://ax.dev/tutorials/gpei_hartmann_service.html#Special-Cases) of the tutorial). It's the exact same thing that @Balandat is suggesting, just using `AxClient` in case that is a more familiar interface.

You'd end up with something like this:
```
ax_client.create_experiment(...)

# Attach all the custom trials and their data
for _ in range(n):
    parameterization, trial_index = ax_client.attach_trial(custom_parameterization)
    ax_client.complete_trial(trial_index=trial_index, raw_data=(metric_mean, metric_SEM))

# Instantiate the model
model = Models.FULLYBAYESIAN(
    experiment=ax_client.experiment, 
    data=ax_client.experiment.lookup_data(),
    num_samples=256,  # Increasing this may result in better model fits
    warmup_steps=512,  # Increasing this may result in better model fits
    gp_kernel="rbf",  # "rbf" is the default in the paper, but we also support "matern"
    torch_device=tkwargs["device"],
    torch_dtype=tkwargs["dtype"],
    verbose=False,  # Set to True to print stats from MCMC
    disable_progbar=True,  # Set to False to print a progress bar from MCMC
)

# To cross_validation or prediction
model.cross_validate(...)
model.predict(...)
```

**Comment by jrojsel** [comment]
Thank you @Balandat and @lena-kashtelyan! I tried your solution @lena-kashtelyan, and it worked perfectly! 👍 

Full code for future reference:

```python
import pandas as pd
import torch
from ax.modelbridge.registry import Models
from ax.service.ax_client import AxClient

tkwargs = {
    "dtype": torch.double,
    "device": torch.device("cuda" if torch.cuda.is_available() else "cpu"),
}

df = pd.read_excel("results.xlsx")

X_ARRAY = df[df.columns[3:-2]].values  # array of size n x m
X_NAMES = df.columns[3:-2].tolist()
X = [[y for y in x] for x in X_ARRAY]
Y = df["Final eta fraction"].tolist()  # list with floats of size n

custom_parameters = []
for i, name in enumerate(X_NAMES):
    custom_parameters.append(
        {
            "name": name,
            "type": "range",
            "bounds": [float(X_ARRAY[:, i].min()), float(X_ARRAY[:, i].max())],
        }
    )

ax_client = AxClient()
ax_client.create_experiment(parameters=custom_parameters)

for i, y in enumerate(Y):
    parameterization, trial_index = ax_client.attach_trial(
        parameters={name: value for name, value in zip(X_NAMES, X[i])}
    )
    ax_client.complete_trial(trial_index=trial_index, raw_data=(y, None))

model = Models.FULLYBAYESIAN(
    experiment=ax_client.experiment,
    data=ax_client.experiment.lookup_data(),
    num_samples=256,  # Increasing this may result in better model fits
    warmup_steps=512,  # Increasing this may result in better model fits
    gp_kernel="rbf",  # "rbf" is the default in the paper, but we also support "matern"
    torch_device=tkwargs["device"],
    torch_dtype=tkwargs["dtype"],
    verbose=False,  # Set to True to print stats from MCMC
    disable_progbar=False,  # Set to False to print a progress bar from MCMC
)
```

Closing.

**Comment by lena-kashtelyan** [comment]
Thank you for leaving the code here for posterity, @jrojsel!
